Use API to initiate repository installation.

davebx opened the pull request.

davebx committed:
Use API to initiate repository installation.

davebx committed:
Fix imminent import error in travis.

martenson commented:
I like this change! Installation tested locally.Thanks for working on this
:+1:

bgruening commented:
I'm just wondering if there is a better name as `initiate_repository_installation`. Can we kill `initiate_`?

davebx commented:
I'm open to making it `repository_installation`

martenson commented:
`install_repository` ?

bgruening commented:
:+1: for  `install_repository`

dannon commented:
Or even just `install`?  I mean, we're already at /api/tool_shed_repositories/, right?

davebx commented:
:+1: for `install`

nsoranzo commented:
I think it's not very clear the difference between `initiate_repository_installation()` and `install_repository_revision()`.

dannon commented:
@nsoranzo What is the difference?  The docstrings don't make that clear to me either.

jmchilton commented on templates/admin/tool_shed_repository/initiate_repository_installation.mako:
```diff
@@ -17,11 +17,11 @@
             $.ajax( {
                 type: "POST",
-                url: "${h.url_for( controller='admin_toolshed', action='manage_repositories' )}",
+                url: "/api/tool_shed_repositories/initiate_repository_installation",
```
Will this work if Galaxy is served on a prefix - I feel like this was using url_for a reason.

davebx replied:
This is a good point. Someone should fix that.

davebx commented:
@nsoranzo @dannon `install_repository_revision(s)` both wait until the installations complete before returning status, whereas `install` just tells the galaxy process to start installing, while the installation monitoring page updates the status.

dannon commented:
Ahh.  Can we just unify this instead of having two endpoints then, perhaps with an optional 'block' flag or something?

(though I'm having trouble seeing when blocking is a really good idea wrt: web requests like this that might take a while)

davebx commented:
@dannon Yes. As for the blocking, that API method was designed for command line installation of repositories.

davebx committed:
Rename API endpoint. Correct url for API endpoint.

davebx committed:
Typo.

davebx committed:
Additional cleanup.

bgruening commented:
I assume that one reason this was a blocking API is that it is not possible to install multiple repositories at the same time. What about adding a warning message about this into the API?

dannon commented:
I'm not sure about the origin story here (and probably don't want to look ;)), but in any event a web request should basically _never_ block unless it's a < 500ms (and that's being _really_ generous) operation, API or not.

This should return a handle to an ongoing install process (which is really just the tool identifier, since everything else is derivable from there) and command line tools (and anything else) should poll for status.

davebx commented:
@dannon I like that idea. And then the endpoint for installation status updates could potentially be something like `/repositories/install/{status identifier}/status`

Although I'm not sure the tool identifier, unless it's a GUID, would be specific enough. I'll give it some thought, maybe bother @martenson about it, and get back to you.

dannon commented:
@davebx Yep, I assumed the full tool guid.

davebx committed:
Add API endpoint to show installation status.

This will return a dict of the repository, with the repository_dependencies
element containing a list of dicts with the dependency chain descending from
that repository.

hexylena commented:
FYI @galaxybot will comment on PRs with Emoji +1s, so use text +1 if you want it to ignore your comments

martenson commented:
@davebx can we get a rebase please?

davebx commented:
This code will be included in my upcoming pull request based on https://github.com/davebx/galaxy/tree/toolshed_decouple
